clcache.py

```diff
@@ -1612,19 +1621,26 @@ def processCompileRequest(cache, compiler, args):
     printOutAndErr(out, err)
     return exitCode
 
+def filterSourceFiles(cmdLine, sourceFiles):
+    # type: (List[str], List[Tuple[str, str]]) -> Iterator[str]
+    setOfSources = set(sourceFile for sourceFile, _ in sourceFiles)
+    skippedArgs = ('/Tc', '/Tp', '-Tp', '-Tc')
+    yield from (
+        arg for arg in cmdLine
+        if not (arg in setOfSources or arg.startswith(skippedArgs))
+    )
+
 def scheduleJobs(cache, compiler, cmdLine, environment, sourceFiles, objectFiles):
-    baseCmdLine = []
-    setOfSources = set(sourceFiles)
-    for arg in cmdLine:
-        if not (arg in setOfSources or arg.startswith("/MP")):
-            baseCmdLine.append(arg)
+    # type: (???, str, List[str], ???, List[Tuple[str, str]], List[str]) -> int
+    # Filter out all source files from the command line to form baseCmdLine
+    baseCmdLine = [arg for arg in filterSourceFiles(cmdLine, sourceFiles) if arg != '/MP']
```

Comment:
The `arg != /MP` test is what breaks the `TestRunParallel.testHitsViaMpConcurrent` test.

The test first compiles two source files individually using the command lines

1. `/nologo /EHsc /c fibonacci01.cpp`
2. `/nologo /EHsc /c fibonacci02.cpp`

The test then tries to compile both files again via

* `/nologo /EHsc /c /MP2 fibonacci01.cpp fibonacci02.cpp`

...expecting that this triggers two cache hits. This causes clcache to reinvoke itself recursively for each of the source files. However, due to the `arg != '/MP'` test, the recursive invocations still include the `/MP2` flag, i.e. the two nested instances use the command lines

1. `/nologo /EHsc /c /MP2 fibonacci01.cpp`
2. `/nologo /EHsc /c /MP2 fibonacci02.cpp`

Thus, the command line are different than they were when the files were originally compiled, hence the hash sum for the cache entry is different and hence there is no cache hit.

It seems this can be fixed by using `not args.startswith('/MP')` instead of `arg != '/MP'`.